### DISCUSSION TITLE
Debug/Build Multiple Projects in same workspace

### DISCUSSION THREAD

**Discussion by angelhdzmultimedia** [BowlerHatLLC/vscode-as3mxml#739] [Q&A] [answered]
Hello!

Have this in the root of the workspace:

`.vscode/launch.json`
```json
  {
      "type": "swf",
      "request": "launch",
      "name": "Launch Server",
      "internalConsoleOptions": "openOnSessionStart",
      "preLaunchTask": "Build Debug AIR Server",
      "program": "${workspaceFolder}/server/bin/Server.swf"
    },
    {
      "type": "swf",
      "request": "launch",
      "name": "Launch Client",
      "internalConsoleOptions": "openOnSessionStart",
      "preLaunchTask": "Build Debug AIR Client",
      "program": "${workspaceFolder}/client/bin/Client.swf"
    }
```

`.vscode/tasks.json`

```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "actionscript",
      "debug": true,
      "air": "air",
      "asconfig": "${workspaceFolder}/server/asconfig.json",
      "group": "build",
      "problemMatcher": [],
      "label": "Build Debug AIR Server"
    },
    {
      "type": "actionscript",
      "debug": true,
      "air": "air",
      "asconfig": "${workspaceFolder}/client/asconfig.json",
      "group": "build",
      "problemMatcher": [],
      "label": "Build Debug AIR Client"
    }
   
  ]
}
```

And when debugging there are errors no matter the combination of settings tried.

Have a `client` and `server` dirs each with its own src, asconfig.json, application descriptor .xml, etc.
and a `libs` dir where all the shared actionscript packages between the client and server are placed.

![image](https://github.com/BowlerHatLLC/vscode-as3mxml/assets/10191085/84f56cb0-db31-4857-b895-5646cabf5e69)

Any tip about this will be highly appreciated.

**Comment by angelhdzmultimedia** [comment]
### Update
For a moment, abandoned the whole root thing, and opened the server dir in VSCode, and added its own launch and tasks configurations and it worked...

So tried to copy that configuration back to the root thing,  abandoned the tasks.json, and I have this launch config now:

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "swf",
      "request": "launch",
      "name": "Launch Server",
      "internalConsoleOptions": "openOnSessionStart",
      "rootDirectory": "server",
      "asconfigPath": "server/asconfig.json",
      "program": "server/src/Server-app.xml"
    },
    {
      "type": "swf",
      "request": "launch",
      "name": "Launch Client",
      "internalConsoleOptions": "openOnSessionStart",
      "rootDirectory": "client",
      "asconfigPath": "client/asconfig.json",
      "program": "client/src/Client-app.xml"
    }
  ]
}
```

Still not working...

**Comment by joshtynjala** [comment, marked as answer]
In _launch.json_, you want to point to the `*-app.xml` files in your output directories.

```json
{
	"version": "0.2.0",
	"configurations": [
		{
			"type": "swf",
			"request": "launch",
			"name": "Launch Server",
			"program": "server/bin/Server-app.xml"
		},
		{
			"type": "swf",
			"request": "launch",
			"name": "Launch Client",
			"program": "client/bin/Client-app.xml"
		}
	]
}
```

**Reply by angelhdzmultimedia** [reply]
Wow! Of all the multiple combinations tried, never thought of using the descriptor from  `bin` instead of the one from `src`! 🔥🤯👏

Good catch! Thank you!

**Reply by angelhdzmultimedia** [reply]
How to set different ports for each config?

I get a "address in use" with one of the two, and the other debugs normally. Sometimes it's `Launch Client` and sometimes `Launch Server` interchangeably.

![image](https://github.com/BowlerHatLLC/vscode-as3mxml/assets/10191085/bd5ab756-8747-4385-a757-3d624808f027)

**Reply by joshtynjala** [reply]
I don't think that I've ever tried launching two debug sessions at the same time. I'm not sure that there is a way for AIR to do that. You may need to run one session without debugging.

**Reply by angelhdzmultimedia** [reply]
### Forgot to mention!
I'm using **compounds**.

![image](https://github.com/BowlerHatLLC/vscode-as3mxml/assets/10191085/b07360d1-06c1-4367-9edb-d210b7dd5f2c)

With compounds, I get the error mentioned earlier, but if I first debug `Launch Server` and then `Launch Client`, it works! LOL 😂👀

![image](https://github.com/BowlerHatLLC/vscode-as3mxml/assets/10191085/a7ca0465-21f3-4e88-8722-82b137c61b2a)

**Reply by angelhdzmultimedia** [reply]
### Solved

You're right hahaha! I set a task for the server launch config and set debug false and now they both work! One with debug and the other without debu! Genius! 🤯🔥👏😍

**Reply by joshtynjala** [reply]
I wasn't familiar with compounds. Interesting! Makes sense for when you need to launch multiple things at the same time.

If I had to guess, AIR might try to find an open port as part of the debugging initialization process, and if you launch two AIR instances at nearly the same time, they both try to connect to the same open port. One instance claims it first, but the other doesn't know, and tries to claim it too. If you manually launch both, that might be enough time passing that the second instance correctly identifies that the first instance claimed the port, and it is not open, so it tries a different port instead.

**Reply by angelhdzmultimedia** [reply]
I've done it haha! 🤯🔥😍👏😂🎉

Added a `preLaunchTask` to the `Launch Client` debug config, that executes a python script, which checks every 1 second for a window with title `"Server"` and then it continues execution. So when I call the compound, it runs `Launch Server`, then waits, and then runs `Launch Client` when the Server.swf ADT window is open.

♻️I hate programming. ♻️ I love programming. Repeat♻️. Hahaha!!! 🥲🤷‍♂️

I will keep trying to find a way to give different ports to ADT/FDB though. 👀🔎

Thanks for everything again!
